service/storage_proxy.cc

```diff
@@ -3581,8 +3600,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             return _proxy->query_result_local(_schema, _cmd, _partition_range, opts, _trace_state, timeout);
         } else {
             tracing::trace(_trace_state, "read_data: sending a message to /{}", ep);
-            return ser::storage_proxy_rpc_verbs::send_read_data(&_proxy->_messaging, netw::messaging_service::msg_addr{ep, 0}, timeout, *_cmd, _partition_range, opts.digest_algo).then([this, ep](rpc::tuple<query::result, rpc::optional<cache_temperature>> result_hit_rate) {
-                auto&& [result, hit_rate] = result_hit_rate;
+            return ser::storage_proxy_rpc_verbs::send_read_data(&_proxy->_messaging, netw::messaging_service::msg_addr{ep, 0}, timeout, *_cmd, _partition_range, opts.digest_algo).then([this, ep](rpc::tuple<query::result, rpc::optional<cache_temperature>, rpc::optional<std::optional<replica::exception_variant>>> result_hit_rate) {
+                auto&& [result, hit_rate, opt_exception] = result_hit_rate;
+                if (opt_exception.has_value() && opt_exception->has_value()) {
```

Comment:
Somewhat sad but can only be fixed in rpc. @gleb-cloudius we should have a way to pass default-constructible optional values, so that if they're missing they are replaced by a default-constructed value rather than an optional.

Reply:
You can do it in out IDL for new members in structs since there is a syntax for it (you can even specify default value). For types that are directly managed by rpc it is not clear how to specify what you want. May be new rpc type rpc::optional_with_default_value(). Not sure it worth it to remove the `if`.

I am also not sure why do we need `std::optional` inside `rpc::optional` here. Doesn't `replica::exception_variant` have "empty" value?

Reply:
We could just say that if the message ends before an argument is deserialized, it is default-constructed (if possible). Just like we say it for structs.

Reply:
And if it is not? Fail compilation? Or do different things in compile time based on if `T` is default constructable? But then rpc may change without notice by changing T without doing any IDL changes. Also some code may want to know if a value was provided even if `T` is default construcatble and there should be a way to express it.

IIRC for structs we require to specify default value in IDL to not get a value as rpc::optional.

Reply:
Also RPC has to know what arguments are optional because it needs to be able to detect that data is missing. Suppose we have an RPC call `foo(bool, bool)`. If a fragment has only enough data for one `bool` we should signal underflow. OTOH if we had `bar(bool)` and later extended it to have `bar(bool, bool)` having only one `bool` in a networking frame is not an error.

Reply:
How about this:
- new arguments have to be wrapped in rpc::optional in IDL
- we provide three options for the C++ handler signature:
   1. std::optional (disengaged if argument not provided)
   2. initializer in IDL (used to initialize if argument not provided)
   3. default constructor (requires that T is default constructible)
  
IIUC these changes would all be in IDL, not RPC.